core/protocols/aws-xml-protocol/src/main/java/software/amazon/awssdk/protocols/xml/internal/unmarshall/AwsXmlPredicatedResponseHandler.java

```diff
@@ -0,0 +1,183 @@
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.protocols.xml.internal.unmarshall;
+
+import java.util.Optional;
+import java.util.function.Function;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.Response;
+import software.amazon.awssdk.core.SdkPojo;
+import software.amazon.awssdk.core.SdkStandardLogger;
+import software.amazon.awssdk.core.exception.RetryableException;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.core.exception.SdkException;
+import software.amazon.awssdk.core.http.HttpResponseHandler;
+import software.amazon.awssdk.core.interceptor.ExecutionAttributes;
+import software.amazon.awssdk.http.SdkHttpFullResponse;
+import software.amazon.awssdk.protocols.query.unmarshall.XmlElement;
+import software.amazon.awssdk.utils.IoUtils;
+
+/**
+ * Unmarshalls an HTTP response into either a successful response POJO, or into a (possibly modeled) exception based
+ * on a predicate that the unmarshalled response can be tested against. Returns a wrapper {@link Response} object which
+ * may contain either the unmarshalled success POJO, or the unmarshalled exception.
+ *
+ * @param <OutputT> Type of successful unmarshalled POJO.
+ */
+@SdkInternalApi
+public class AwsXmlPredicatedResponseHandler<OutputT> implements HttpResponseHandler<Response<OutputT>> {
+    private static final Logger log = LoggerFactory.getLogger(AwsXmlPredicatedResponseHandler.class);
+
+    private final Function<SdkHttpFullResponse, SdkPojo> pojoSupplier;
+    private final Function<AwsXmlUnmarshallingContext, OutputT> successResponseTransformer;
+    private final Function<AwsXmlUnmarshallingContext, ? extends SdkException> errorResponseTransformer;
+    private final Function<AwsXmlUnmarshallingContext, AwsXmlUnmarshallingContext> decorateContextWithError;
+    private final boolean needsConnectionLeftOpen;
+
+    /**
+     * Standard constructor
+     * @param pojoSupplier A method that supplies an empty builder of the correct type
+     * @param successResponseTransformer A function that can unmarshall a response object from parsed XML
+     * @param errorResponseTransformer A function that can unmarshall an exception object from parsed XML
+     * @param decorateContextWithError A function that determines if the response was an error or not
+     * @param needsConnectionLeftOpen true if the underlying connection should not be closed once parsed
+     */
+    public AwsXmlPredicatedResponseHandler(
+        Function<SdkHttpFullResponse, SdkPojo> pojoSupplier,
```

Comment:
Not sure about the constructor with 5 parameters and 4 of them are functions. My concern is that we might add more fields in the future and it's going to be super difficult to maintain different combinations of ctors.

Reply:
It's internal and only called from one place. I propose we keep it as-is for now and refactor in future if necessary as we're free to do so.